widgetry/src/backend_glow.rs

```diff
@@ -418,180 +476,170 @@ impl PrerenderInnards {
-    // store textures in a grid within a single spritesheet image, where a row and column traverses
-    // multiple sprites.
-    //
-    // For example, if we had 6 textures, A-F, the input spritesheet bytes would be like:
-    // [[AAA, BBB, CCC],
-    //  [AAA, BBB, CCC]
-    //  [AAA, BBB, CCC],
-    //  [DDD, EEE, FFF],
-    //  [DDD, EEE, FFF],
-    //  [DDD, EEE, FFF]]
-    //
-    // Which we need to convert to:
-    // [[AAAAAAAAA],
-    //  [BBBBBBBBB],
-    //  [CCCCCCCCC],
-    //  [DDDDDDDDD],
-    //  [EEEEEEEEE],
-    //  [FFFFFFFFF]]
-    use image::GenericImageView;
-    for y in 0..sprites_per_column {
-        for x in 0..sprites_per_row {
-            let sprite_cell = img.view(
-                x * sprite_length,
-                y * sprite_height,
-                sprite_length,
-                sprite_height,
-            );
-            for p in sprite_cell.pixels() {
-                formatted_pixel_bytes.extend_from_slice(&p.2 .0);
+        info!(
+            "img_size: {}x{}px ({} px), sprite_size: {}x{}px, sprites: {}x{} ({} sprites)",
+            img_width,
+            img_height,
+            img.pixels().len(),
+            sprite_width,
+            sprite_height,
+            sprites_per_row,
+            sprites_per_column,
+            sprite_count
+        );
+
+        let mut texture_bytes: Vec<u8> = Vec::with_capacity(img.pixels().len() * bytes_per_pixel);
+
+        // In order to avoid branching in our shader logic, all shapes are rendered with a texture.
+        // Even "non-textured" styles like Fill::Color, use a "default" no-op (pure white) texture,
+        // which we generate here.
+        texture_bytes.append(&mut vec![
+            255;
+            (sprite_width * sprite_height) as usize
+                * bytes_per_pixel
+        ]);
+
+        use image::GenericImageView;
+        for y in 0..sprites_per_column {
+            for x in 0..sprites_per_row {
+                let sprite_cell = img.view(
+                    x * sprite_width,
+                    y * sprite_height,
+                    sprite_width,
+                    sprite_height,
+                );
+                for p in sprite_cell.pixels() {
+                    texture_bytes.extend_from_slice(&p.2 .0);
+                }
             }
         }
-    }
 
-    // prepare and generate mipmaps
-    unsafe {
-        gl.tex_sub_image_3d(
-            target,
-            0,
-            0,
-            0,
-            0,
-            sprite_length as i32,
-            sprite_height as i32,
-            sprite_count as i32,
-            format,
-            glow::UNSIGNED_BYTE,
-            glow::PixelUnpackData::Slice(&formatted_pixel_bytes),
-        );
+        Ok(Self {
+            texture_bytes,
+            sprite_width,
+            sprite_height,
+            sprite_count,
+        })
+    }
 
-        gl.tex_image_3d(
```

Comment:
Turned these next three calls into a loop - I think it's easier to see what it's doing that way.